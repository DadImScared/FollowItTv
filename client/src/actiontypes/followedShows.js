
export const FOLLOW_SHOW = 'followedShows/FOLLOW_SHOW';
export const UNFOLLOW_SHOW = 'followedShows/UNFOLLOW_SHOW';
export const ADD_DAY = 'followedShows/ADD_DAY';
