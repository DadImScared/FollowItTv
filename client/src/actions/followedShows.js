
import _ from 'lodash';
import axios from 'axios';

import axiosOptions from './axiosOptions';
import * as followedActionTypes from '../actiontypes/followedShows';
import { addShows } from './shows';

export const addDay = (day, shows) => {
  return {
    type: followedActionTypes.ADD_DAY,
    day,
    shows
  };
};

export const followShow = (days, id) => {
  return {
    type: followedActionTypes.FOLLOW_SHOW,
    days,
    id
  };
};

export const unfollowShow = (days, id) => {
  return {
    type: followedActionTypes.UNFOLLOW_SHOW,
    days,
    id
  };
};

export const postFollow = async (showId) => {
  return await axios.post(`/api/v1/followshow/${showId}/`, {}, axiosOptions());
};

export const addShowIds = (showIdList) => {
  return {
    type: followedActionTypes.ADD_SHOW_IDS,
    showIdList
  };
};

export const extractShowData = (data) => {
  let isArrayOfNumbers;
  const showIds = new Set();
  const showData = {};
  if (typeof data[0] === 'number') {
    isArrayOfNumbers = true;
  }
  else {
    data.forEach((show) => {
      showIds.add(show.id);
      showData[show.id] = show;
    });
  }
  return {
    showData,
    showIds: isArrayOfNumbers ? data:[...showIds]
  };
};

export const getFollowedShows = async (dispatch, day = '') => {
  try {
    const { data } = await axios.get(`/api/v1/followshows/${day ? `${_.capitalize(day)}/`:''}`, axiosOptions());
    const { showData, showIds } = extractShowData(data);
    if (day) {
      dispatch(addShows(showData));
    }
    dispatch(addShowIds(showIds));
  }
  catch (e) {
    throw e;
  }
};
