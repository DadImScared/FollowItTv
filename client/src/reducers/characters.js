
import { ADD_CHARACTERS } from '../actiontypes/characters';

function characters(state = {}, action) {
  switch(action.type) {
    case ADD_CHARACTERS:
      return {
        ...state,
        ...action.characters
      };
    default:
      return state;
  }
}

export default characters;
