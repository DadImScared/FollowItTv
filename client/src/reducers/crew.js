
import { ADD_CREW } from '../actiontypes/crew';

function crew(state = {}, action) {
  switch (action.type) {
    case ADD_CREW:
      return {
        ...state,
        [action.id]: {
          ...state[action.id],
          ...action.crewMembers
        }
      };
    default:
      return state;
  }
}

export default crew;
