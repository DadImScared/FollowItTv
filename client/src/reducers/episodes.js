
import { ADD_EPISODES } from '../actiontypes/episodes';


function episodes(state={}, action) {
  switch (action.type) {
  case ADD_EPISODES:
    return {
      ...state,
      ...action.episodes
    };
  default:
    return state;
  }
}

export const getEpisodes = (state) => state.episodes;

export default episodes;
