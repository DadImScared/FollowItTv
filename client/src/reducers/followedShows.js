
import _ from 'lodash';
import { createSelector } from 'reselect';

import * as followedShowsTypes from '../actiontypes/followedShows';
import { followShow, unfollowShow } from './followedShowsUtility';
import { getLoading } from './loading';
import { getShows } from './shows';


const updateShowInDay = (dayState, action, updateCb)=> {
  const updatedDays = action.days.filter((day) => day in dayState).map((day) => {
    return {
      [day]: updateCb(dayState[day], action)
    };
  }).reduce((obj, item) => ({ ...obj, ...item }), {});
  return {
    ...dayState,
    ...updatedDays
  };
};

function followedShows(state={}, action) {
  switch(action.type) {
  case followedShowsTypes.ADD_DAY:
    return {
      ...state,
      [action.day]: action.shows
    };
  case followedShowsTypes.UNFOLLOW_SHOW:
    return updateShowInDay(state, action, unfollowShow);
  case followedShowsTypes.FOLLOW_SHOW:
    return updateShowInDay(state, action, followShow);
  default:
    return state;
  }
}

const getFollowedShowsByDay = (state, { day }) => {
  return !day || day === 'All' ? state.followedShowsById : state.followedShows[day] || [];
};

export const getFollowedShows = createSelector(
  [getFollowedShowsByDay, getShows],
  (followedShows, shows) => _.sortBy(followedShows.map((id) => shows[id]), ['name'])
);

export const makeFollowedShowData = () => {
  return createSelector(
    [getFollowedShows, getLoading('MY_SHOWS_', 'day', 'All')],
    (followedShows, isLoading) => ({ followedShows, isLoading, isCached: followedShows.length  })
  );
};

export default followedShows;
