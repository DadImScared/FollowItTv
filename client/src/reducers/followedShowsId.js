
import * as followedShowsTypes from '../actiontypes/followedShows';

import { followShow, unfollowShow } from './followedShowsUtility';

function followedShowsId(state=[], action) {
  switch(action.type) {
  case followedShowsTypes.FOLLOW_SHOW:
    return followShow(state, action);
  case followedShowsTypes.UNFOLLOW_SHOW:
    return unfollowShow(state, action);
  default:
    return state;
  }
}

export default followedShowsId;
