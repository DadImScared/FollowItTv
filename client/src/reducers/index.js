
import { combineReducers } from 'redux';

import followedShows from './followedShows';
import followedShowsById from './followedShowsId';
import shows from './shows';
import schedule from './schedule';
import users from './users';
import episodes from './episodes';
import seasons from './seasons';
import search from './search';

const reducer = combineReducers({
  followedShows,
  followedShowsById,
  seasons,
  episodes,
  shows,
  schedule,
  users,
  search
});

export default reducer;
