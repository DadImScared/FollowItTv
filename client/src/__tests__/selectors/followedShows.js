
import { makeFollowedShowData } from '../../reducers/followedShows';

describe('selectors', () => {
  test('makeFollowedShowData with no day', () => {
    const state = {
      shows: { 1: { name: 'show1' } },
      followedShowsById: [1],
      loading: {
        'MY_SHOWS_All': true
      },
      followedShows: {
        'Wednesday': [1]
      }
    };
    const followedShows = makeFollowedShowData();
    followedShows(state, {});
    followedShows(state, {});
    expect(followedShows.recomputations()).toEqual(1);
    expect(followedShows(state, {})).toMatchSnapshot();
  });
});
