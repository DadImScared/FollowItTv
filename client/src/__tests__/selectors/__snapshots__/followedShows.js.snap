// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors makeFollowedShowData with no day 1`] = `
Object {
  "followedShows": Array [
    Object {
      "name": "show1",
    },
  ],
  "isCached": 1,
  "isLoading": true,
}
`;
